infra/Pos/Communication/Relay/Logic.hs

```diff
@@ -361,102 +392,98 @@ data InvReqDataFlowLog =
-       , Bi (InvOrData key contents)
-       , Bi (ReqMsg key)
-       )
-    => Text -> EnqueueMsg m -> Msg -> key -> contents -> m ()
-invReqDataFlowNeighbors what enqueue msg key dt = handleAll handleE $
-    void $ enqueue msg (\addr _ -> pure (Conversation (invReqDataFlowDo what key dt addr))) >>= waitForConversations
-  where
-    handleE e = logWarning $
-        sformat ("Error sending "%stext%", key = "%build%" to neighbors: "%shown) what key e
-
+-- | Do an Inv/Req/Data/Res conversation (peers determined by the 'EnqueueMsg m'
+-- argument) and wait for the results.
+-- This will wait for all conversations to finish. Exceptions in the conversations
+-- themselves are caught and returned as Left. If the peer did not ask for the
+-- data, then Right Nothing is given, otherwise their response to the data is
+-- given.
 invReqDataFlow
     :: forall key contents m.
        ( Message (InvOrData key contents)
-       , Message (ReqMsg key)
+       , Message (ReqOrRes key)
        , Bi (InvOrData key contents)
-       , Bi (ReqMsg key)
+       , Bi (ReqOrRes key)
        , Buildable key
        , MinRelayWorkMode m
        , MonadGState m
        )
-    => Text -> EnqueueMsg m -> Msg -> key -> contents -> m ()
+    => Text
+    -> EnqueueMsg m
+    -> Msg
+    -> key
+    -> contents
+    -> m (Map NodeId (Either SomeException (Maybe (ResMsg key))))
 invReqDataFlow what enqueue msg key dt = handleAll handleE $ do
     its <- enqueue msg $
         \addr _ -> pure $ Conversation $ invReqDataFlowDo what key dt addr
-    void $ waitForConversations its
+    waitForConversations (fmap try its)
   where
-    handleE e = logWarning $
-        sformat ("Error sending "%stext%", key = "%build%": "%shown)
+    handleE e = do
+        logWarning $
+            sformat ("Error sending "%stext%", key = "%build%": "%shown)
                 what key e
+        throw e
 
 invReqDataFlowDo
-    :: ( Message (InvOrData key contents)
-       , Message (ReqMsg key)
-       , Bi (InvOrData key contents)
-       , Bi (ReqMsg key)
-       , Buildable key
+    :: ( Buildable key
        , MinRelayWorkMode m
        , MonadGState m
        )
     => Text
     -> key
     -> contents
     -> NodeId
-    -> ConversationActions (InvOrData key contents) (ReqMsg key) m
-    -> m ()
+    -> ConversationActions (InvOrData key contents) (ReqOrRes key) m
+    -> m (Maybe (ResMsg key))
 invReqDataFlowDo what key dt peer conv = do
     send conv $ Left $ InvMsg key
-    recvLimited conv >>= maybe handleD replyWithData
+    it <- recvLimited conv
+    maybe (handleD logDebug >> return Nothing) replyWithData it
   where
     -- TODO need to check we're asked for same key we have
-    replyWithData (ReqMsg _) = send conv $ Right $ DataMsg dt
-    handleD =
-        logDebug $
+    replyWithData (Left (ReqMsg _)) = do
+        send conv $ Right $ DataMsg dt
+        it <- recvLimited conv
+        maybe (handleD logError >> throw UnexpectedEnd) checkResponse it
```

Comment:
There's an existing TODO on line 467 (457 before changes) `need to check we're asked for same key we have`. You'd think that after an `InvMsg` the key doesn't need to be sent back, just assume that they want what we offered. But `ReqMsg` is re-used in a different context: you can start a conversation with a `ReqMsg`, in which case it needs to include the key so the peer knows what you want.